x/wasm/IBC.md

```diff
@@ -178,96 +178,98 @@ type IBCCloseChannel struct {
+}
 
-```go
-package lifecycle
-
-// if this returns error, we reject the channel opening
-// otherwise response has the versions we accept
-//
-// It is provided the full ChannelInfo being proposed to do any needed checks
-func OnChanNegotiate(ctx sdk.Context, k *wasm.Keeper, contractAddress sdk.AccAddress, request ChannelInfo) 
-    (version Version, err error) {}
-
-// This is called with the full ChannelInfo once the other side has agreed.
-// An error here will still abort the handshake process. (so you can double check the order/version here)
-//
-// The main purpose is to allow the contract to set up any internal state knowing the channel was established,
-// and keep a registry with that ChannelID
-func OnChanOpened(ctx sdk.Context, k *wasm.Keeper, contractAddress sdk.AccAddress, request ChannelInfo) error {}
-
-// This is called when the channel is closed for any reason
-// TODO: any meaning to return an error here? we cannot abort closing a channel
-func OnChanClosed(ctx sdk.Context, k *wasm.Keeper, contractAddress sdk.AccAddress, request ChannelClosedInfo) error {}
-
-type ChannelInfo struct {
-    // key info to enforce (error if not what is expected)
-    Order channeltypes.Order
-    // The proposed version. This may come from the relayer, from the initiating contract, or the responding contract.
-    // In any case, this is the currently agreed upon version to use and if there is disagreement, the contract should
-    // propose another one (if possible in return value), or return an error
-    ProposedVersion Version
-    // local id for the Channel that is being initiated
-    ChannelID string
-    // these two are taken from channeltypes.Counterparty
-    RemotePortID string
-    RemoteChannelID string
+type IBCPacketTimeoutResponse struct {
+  Messages []sdk.Msg `json:"messages"`
+  Log []wasmTypes.LogAttribute `json:"log"`
 }
 
-type ChannelClosedInfo struct {
-    // local id for the Channel that is being shut down
-    ChannelID string
+type IBCChannelOpenResponse struct {
+  // Result contains a boolean if the channel would be accepted
+  Result                       bool   `json:"result"`
+  // Reason optional description why it was not accepted
+  Reason                       string `json:"reason"`
+}
+
+type IBCChannelConnectResponse struct {
+  Messages []sdk.Msg `json:"messages"`
+  Log []wasmTypes.LogAttribute `json:"log"`
+}
+
+type IBCChannelCloseResponse struct {
+  Messages []sdk.Msg `json:"messages"`
+  Log []wasmTypes.LogAttribute `json:"log"`
+}
+
+// Environment data for a contract
+type Env struct {
+...
+  // optional IBC meta data only set when called in IBC context
+  IBC *IBCInfo `json:"ibc,omitempty"`
 }
+
+type IBCInfo struct {
+  // Port of the contract
+  Port string `json:"port"`
+  // Channel to the contract
+  Channel string         `json:"channel"`
+  // Optional packet meta data when called on IBC packet functions
+  Packet  *IBCPacketInfo `json:"packet,omitempty"`
```

Comment:
3 levels Env -> IBCInfo -> IBCPacketInfo seems overly complex. I had a much simpler version that covers most of this.

Let's flatten it, and we don't need both (Port, Channel) and (SourcePort, SourceChannel). Just show the local side and they can use `GetCounterParty` if they need it. Let's trim this to the essential

Reply:
Okay, this is passed on all calls, connections as well as packets, which is why you have the PacketInfo optional.

This is getting confusing. I would like things always there or not - clear at compile time - not based on out-of-bands understanding

Maybe just wrap the Packet Info around the Msg